### PR TITLE
osd: don't crash on empty snapset

### PR DESCRIPTION
**Pull request by trociny**
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

### CODE REVIEW

**Comment by trociny** [comment]
@liewegas For one of our customers we observed osd crashes due to unexpectedly empty snapset returned for some objects in the cache tire pool. And the proposed patch helped to make their cluster functional again.

This looks like related to https://tracker.ceph.com/issues/21557, and although the root cause is unknown, wouldn't it be a good idea to have these guards upstream so this type of inconsistency would not make the cluster down? Or do you have other suggestions how it could be addressed?

**Comment by liewegas** [comment]
I think the best would be to assert if the debug config option is true so that we can catch it qa.  Although we have failed to do that.. unclear what the root cause is :(

See https://github.com/ceph/ceph/pull/20040.  Ideally we'd make some attempt to fix the inconsistency during scrub...

**Comment by trociny** [comment]
@liewegas 

> I think the best would be to assert if the debug config option is true so that we can catch it qa. 

Thank you! Updated. I cherry-picked your 618f549 from #20040 to make it build. I will rebase when #20040 is merged.

> Ideally we'd make some attempt to fix the inconsistency during scrub...

Interesting. I think it could done as a separate PR. Do you have an idea what info could be used to check/restore a snapset?

**Comment by trociny** [comment]
Rebased after #20040 is merged.

File: src/osd/SnapMapper.cc [hunk 1]
```diff
@@ -155,7 +155,10 @@ int SnapMapper::get_snaps(
     bufferlist::iterator bp = got.begin()->second.begin();
     decode(*out, bp);
     dout(20) << __func__ << " " << oid << " " << out->snaps << dendl;
-    assert(!out->snaps.empty());
+    if (out->snaps.empty()) {
+      dout(1) << __func__ << " " << oid << " empty snapset" << dendl;
+      assert(!cct->_conf->osd_debug_verify_snaps);
```
Comment (smithfarm) [inline src/osd/SnapMapper.cc]: no `return -ENOENT` here?

Comment (trociny) [reply src/osd/SnapMapper.cc]: Yes, it was intentional: we don't want `SnapMapper::get_snaps` to fail here, just return an empty snapset.

**Comment by liewegas** [comment]
> Interesting. I think it could done as a separate PR. Do you have an idea what info could be used to check/restore a snapset?

It might not require anything special, actually: perhaps an empty SnapSet 
will result in any/all clones getting removed (as stray clones) and we'd 
be done with it.  I would give it a try by injecting the corruption 
(perhaps removing it directly via the fuse mountpoint or via 
ceph-objectstore-tool) to verify the io path now behaves and also that 
scrub will too.

**Comment by trociny** [comment]
@liewegas I have temporary added a DNM commit to this PR that shows how I tested this.

The test contains a small tool to clear the snapset in an object snaps blob. It is used together with ceph-objectstore-tool to inject the empty snapset inconsistency. The objects belong to an rbd image with snapshots. I tested that after injecting the corruption `rbd export` still returns the same result. The "empty snapset" message (or osd_debug_verify_snaps assert) is triggered on scrub in `SnapMapper::get_snaps`.

I have failed to trigger "empty snapset" in `PrimaryLogPG::find_object_context()` because I don't know how to reproduce this. Though we saw this in the field [1].

[1] https://tracker.ceph.com/issues/22672

**Comment by trociny** [comment]
@liewegas ping

**Comment by liewegas** [comment]
Looks good to me; let's drop teh DNM commit and then queue for testing?

**Comment by trociny** [comment]
@liewegas thanks! Rebased.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-04-23_07:38:14-rados-wip-kefu-testing-2018-04-23-1357-distro-basic-smithi/

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
@jdurgin @liewegas shall we include this change in mimic?

**Pull request merged**